react-components/src/components/NodeCacheProvider/RevisionFdmNodeCache.ts

```diff
@@ -0,0 +1,232 @@
+    nodes.forEach((n) => {
+      this._treeIndexToFdmData.set(n.treeIndex, nodeData);
+    });
+  }
+
+  private async getDataWithViewsForFdmEdges(
+    nodeEdges: FdmEdgeWithNode[],
+    ancestorsWithSameMapping: Node3D[]
+  ): Promise<Fdm3dNodeData[]> {
+    const nodeInspectionResults = await inspectNodes(
+      this._fdmClient,
+      nodeEdges.map((edge) => edge.edge.startNode)
+    );
+
+    const dataWithViews = nodeEdges.map((fdmEdgeWithNode, ind) => ({
+      fdmId: fdmEdgeWithNode.edge.startNode,
+      view: nodeInspectionResults.items[ind].inspectionResults.involvedViewsAndContainers.views[0],
+      cadNode: fdmEdgeWithNode.node
+    }));
+
+    ancestorsWithSameMapping.forEach((ancestor) =>
+      this._treeIndexToFdmData.set(ancestor.treeIndex, dataWithViews)
+    );
+
+    return dataWithViews;
+  }
+
+  private async getClosestParentMapping(treeIndex: number): Promise<{
+    edges: FdmCadEdge[];
+    ancestorsWithSameMapping: Node3D[];
+    firstMappedAncestorTreeIndex: number;
+  }> {
+    const ancestors: Node3D[] = await fetchAncestorNodesForTreeIndex(
+      this._modelId,
+      this._revisionId,
+      treeIndex,
+      this._cogniteClient
+    );
+
+    const ancestorMappings = await this.getMappingEdgesForAncestors(ancestors);
+
+    if (ancestorMappings.length === 0) {
+      return { edges: [], ancestorsWithSameMapping: [], firstMappedAncestorTreeIndex: 0 };
+    }
+
+    const edgesWithCorrespondingTreeIndex = this.combineEdgesWithTreeIndex(
+      ancestorMappings,
+      ancestors
+    );
+
+    const firstMappedAncestorTreeIndex = findLargestTreeIndex(edgesWithCorrespondingTreeIndex);
+    return getAncestorDataForTreeIndex(
+      firstMappedAncestorTreeIndex,
+      edgesWithCorrespondingTreeIndex,
+      ancestors
+    );
+  }
+
+  private combineEdgesWithTreeIndex(
+    mappingEdges: FdmCadEdge[],
+    nodes: Node3D[]
+  ): Array<{ edge: FdmCadEdge; treeIndex: TreeIndex }> {
+    return mappingEdges.map((edge) => {
+      const ancestorConnectedToEdge = nodes.find((a) => a.id === edge.properties.revisionNodeId);
+
+      assert(ancestorConnectedToEdge !== undefined);
+
+      return {
+        edge,
+        treeIndex: ancestorConnectedToEdge.treeIndex
+      };
+    });
+  }
+
+  private async getMappingEdgesForAncestors(ancestors: Node3D[]): Promise<FdmCadEdge[]> {
+    const cachedFirstMappedAncestor = ancestors
+      .filter((a) => this._treeIndexToFdmEdges.has(a.treeIndex))
+      .sort((a, b) => b.treeIndex - a.treeIndex)[0];
+
+    if (cachedFirstMappedAncestor !== undefined) {
+      const edgesAndNodes = this._treeIndexToFdmEdges.get(cachedFirstMappedAncestor.treeIndex);
+
+      assert(edgesAndNodes !== undefined);
+
+      return edgesAndNodes.map((edge) => edge.edge);
+    }
+
+    const ancestorMappings = await getMappingEdgesForNodeIds(
+      this._modelId,
+      this._revisionId,
+      this._fdmClient,
+      ancestors.map((a) => a.id)
```

Comment:
This is a case where I think `a` is fine as a variable name, since the usage is obvious.

Reply:
Phew :relieved: